### PR TITLE
Translation: per-item language markers, tap-to-toggle, and Tap to Translate mode

### LINKED ISSUE #521: [Feature]: Translated comments or posts should have a “flair” to highlight they have been translated
### Pre-submission checklist

- [x] I searched existing issues and this hasn't already been requested.
- [x] I'm running the latest version of Apollo Reborn, and this feature doesn't already exist.

### What problem does this solve?

I’m using google translation on latest 3.3.0. When a comment is translated it doesn’t shown in any way. I think this is a bit problematic as you may not known that what you are reading has been translated, since translated comments and untranslated ones are exactly identical.

### Describe the feature you'd like

Ideally translated comments should be highlighted somehow.

I suggest to do it similarly to the “deleted” comments. Highlight the translated comments/posts with a proper color (light orange? I would leave yellow out since it’s already used to highlight comments) and put a similarly colored “TRANSLATED” button similar to the “DELETED BY USER” for deleted comments.

There should be a priority logic, since a comment can be both deleted and translated (probably deleted is higher prio, so the comments gets highlighted red, but shows both “button” “TRANSLATED” / “DELETED BY USER”

### Alternatives considered

_No response_

### Was this a previously existing Apollo feature?

No

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by icpryde**
Adds per-item translation indicators to the bulk-translation feature, makes them tappable per-item toggles, and introduces an opt-in **Tap to Translate** mode. Works identically across all providers (Google / LibreTranslate / Apple).

## What's new

**Language markers.** Translated comments get a small italic "🌐 Translated from Spanish" line; feed titles and thread headers get a compact "🌐 PT" marker on the post stats row (matched to the stat font/icon size and baseline-aligned, so it sits level with ↑ 💬 🕐). The code tracks state: it shows the *source* language while translated, and the *target* language while showing the original — so it always names what a tap switches to. **Match App Colour** optionally tints the markers with the app accent instead of green.

**Tap to toggle.** Tapping a marker (or the comment line) flips just that item between translated and original. For posts, the title, body preview and rich link card flip together. Pins survive scrolling, cell reuse, votes and thread reopen; the vote-resilience reapply paths all honour them.

**Tap to Translate mode.** When enabled, nothing auto-swaps: comments show a tappable "🌐 Translate" line and posts show the target-language marker, and tapping translates just that item (tap again to revert). Affordances are driven by on-device language detection, so a slow or failing provider never hides the controls; translations still prefetch in the background so taps are instant. Toggling the mode live-updates everything on screen — no scrolling or relaunch needed.

**Settings.** Three new toggles (Details on Comments & Posts / Details on Titles / Match App Colour, first two default ON) plus Tap to Translate (default OFF). Superseded toggles grey out in the off position (Auto Translate + the Details rows while Tap to Translate is on; everything under the master toggle). Footer rewritten to be shorter.

**Fix.** The proper-noun translation skip now trusts confident language detection (≥ 0.90), so genuinely-foreign Title-Case titles ("Bicampeões de Futsal Masculino") translate instead of being mistaken for names, while detection-fooling strings ("Dua Lipa" → Indonesian at 0.78) stay protected.

**Dev tooling (sim only).** `scripts/apple-bridge.sh` runs a host-side Apple-translation bridge so the Apple provider is fully testable in the simulator (the Translation framework doesn't exist in sim runtimes). Gated behind `APOLLO_SIM_BUILD` — compiled out of device builds.

## Screenshots

| Normal mode: feed markers | Comment "Translated from" | Pinned to original (EN = tap for English) |
|---|---|---|
| feed markers | comment translated from | pinned original |

| Tap to Translate: thread | Tap to Translate: feed | After tap (Apple provider) |
|---|---|---|
| tap mode thread | tap mode feed | tap translated |

## Testing

Extensively sim-verified (iOS 26, glass + classic) with Google and with Apple via the sim bridge: all marker surfaces, both tap directions on comments/titles/headers/link posts, pin survival across scroll/reuse/votes, mid-session mode flips in both directions, and normal-mode regression with the feature toggles off. An adversarial review pass over the new state machine found and fixed 15 edge cases (mode-flip staleness, cell-reuse pin leaks, a wrong-comment restore on recycled cells, thread-safety of the session sets). Device test to follow.

### CODE REVIEW

**Review by JeffreyCA — APPROVED** [review]
Oh nice you got it working w/ Simulator!

**Comment by icpryde** [comment]
Compact mode is now handled. Two fixes pushed:

- **Dots overlap** — in compact layout the ⋯ more-options button sits immediately after the age stat, and the marker's fixed 6pt lead was drawing the globe right on top of it. The marker now anchors just past the dots' trailing edge (`…🕐18h ⋯ 🌐PT`).
- **Edge crowding / wrap** — the compact `PostInfoNode` lays its author + stats rows in a flexWrap horizontal stack, so inline-vs-wrapped is a pure width decision that ignored the out-of-band marker; rows that "just fit" left the globe jammed against the screen edge. The marker's width is now reserved as `spacingAfter` on the ⋯ node, so the stack measures the row as if the marker were a real child — tight rows wrap to the two-line layout like every other post, and the reserved gap after the dots is exactly where the marker lands. Reservation is released on hide; non-compact rows are unchanged.

Verified in the simulator on r/Benfica (compact): fits-with-globe rows stay inline, tight rows wrap cleanly, and rows without a marker are untouched.

**Comment by icpryde** [comment]
One more small polish: tapping a 🌐 marker now fires a light haptic tick as the title toggles, matching the feedback the info-row taps got on #566. Sim-verified the toggle path still works (Equipment ⇄ Equipamentos on r/Benfica).

File: src/ApolloTranslation.xm [hunk 1]
```diff
@@ -2464,6 +2963,20 @@ static NSString *ApolloRichPreviewTranslationCacheKey(NSURL *url, NSString *fiel
 NSString *ApolloRichPreviewTranslatedTextIfAvailable(NSURL *url, NSString *field, NSString *sourceText, id ownerNode) {
     NSString *trimmed = ApolloTrimmedString(sourceText);
     if (trimmed.length < 3) return nil;
+    // Per-post pin: the user tapped this post's feed marker to see the original,
+    // so its card must render untranslated too (toggled back off by another tap).
+    if (sRichPreviewPinnedOriginalURLKeys.count > 0) {
+        NSString *pinKey = ApolloRichPreviewPinKeyForURL(url);
+        if (pinKey.length > 0 && [sRichPreviewPinnedOriginalURLKeys containsObject:pinKey]) return nil;
```
Comment (nickclyde) [inline src/ApolloTranslation.xm]: I think this read of `sRichPreviewPinnedOriginalURLKeys` may be a data race. `ApolloRichPreviewTranslatedTextIfAvailable` runs on ASDK background layout threads (reached via the link-preview rendering path), and right below this the sibling tap-mode set is deliberately gated through the `ApolloTapModeSetsLock` helpers precisely because it's read off-main. But `sRichPreviewPinnedOriginalURLKeys` is a plain `NSMutableSet` mutated on the main thread in `ApolloToggleTranslationForTitleNode` (the lazy-init / `addObject:` / `removeObject:` block) with no synchronization. A concurrent `containsObject:` here during a main-thread mutation of the same set is undefined and can crash or corrupt it. Could we route this set through the same `@synchronized(ApolloTapModeSetsLock())` (or its own lock) on both the read here and the mutations in the toggle, the way the tap-mode sets do?

**Comment by icpryde** [comment]
Pushed a fix for a marker-positioning bug reported after real-world use (commit on this branch).

**Symptom:** the per-post `🌐 XX` feed marker sometimes rendered on its own line under the author row — floating under the username, or overlapping a flair chip — instead of inline after the timestamp. Tapping the marker, or opening the post and coming back, snapped it into place. Only showed up while scrolling ("looked fine in testing").

**Cause:** a race between the marker install and ASDK view mounting. `ApolloUpdatePostInfoMarkerForNode` picks the label's Auto Layout host at call time — the age/time node's view if mounted, else the whole `PostInfoNode` view as a fallback (trailing/centerY, which on a two-row info node lands on the author-row boundary). The feed marker installs from `PostTitleNode`'s preload/display interface-state hooks, and a **cached** translation completes synchronously — so on a scrolled feed the install runs while the cell is still in ASDK's offscreen range, before the age view is mounted, and takes the fallback pin. Nothing re-evaluated the host afterwards (the owned-and-translated early return never reaches the marker updater again). A fresh network translation lands after the cell is visible, which is why first-run testing looked fine.

**Fix:** track whether the label actually anchored to the age view (reusing the previously-unused `kApolloPostInfoMarkerAnchoredKey`), stash the shown source code, and add a `PostInfoNode didEnterVisibleState` hook that re-runs the exact repair the tap path already uses when a marker is still on the fallback. Visible state guarantees the age view is mounted; a single 0.15s retry covers the same-runloop mount race. One-shot and event-driven — no `layoutSubviews` writes.

Verified in the simulator by forcing markers onto the fallback pin (reproducing the exact wrong position) and confirming the repair function re-anchors all of them back inline. Device build clean.

**Comment by icpryde** [comment]
No-regression follow-up: re-ran the built tweak (debug scaffolding removed) on a fresh app launch in the simulator — all feed markers install correctly anchored inline after the stats row (`↑147 💬31 🕐4d ⋯ 🌐EN`) on r/benfica, with the `didEnterVisibleState` re-anchor hook a no-op bail for already-correct markers. Since the hook fires for every `PostInfoNode` (compact rows and the comments-header banner included), worth noting its guard: it only re-runs the updater when a shown marker is still on the fallback pin (or its age view got re-mounted out from under it) — otherwise it returns immediately, so correctly-anchored markers are untouched. Device + simulator builds clean.

**Review by nickclyde — COMMENTED** [review]
One possible thread-safety issue around the new per-post pin set, noted inline.

File: src/ApolloTranslation.xm (hunk 1 above)
Comment (icpryde) [reply src/ApolloTranslation.xm]: Good catch, that read was indeed racing the main-thread mutations. Fixed in 265c003 — went with your first suggestion since the sets are siblings anyway: two helpers next to the existing tap-mode ones (`ApolloRichPreviewPinnedOriginalContainsKey` / `ApolloRichPreviewSetKeyPinnedOriginal`), both `@synchronized(ApolloTapModeSetsLock())`, and the background read here plus all the toggle-path mutations now go through them. No direct access to the set remains outside the two helpers. Re-verified the marker toggle in the sim in both normal and tap-to-translate mode — title + link card still flip both ways.

**Review by nickclyde — APPROVED** [review]
Thanks for addressing, looks good!
